### ISSUE TITLE
Reproducible case that can trigger Lua internal assertion: "results from function overflow current stack size" (LUAI_ASSERT)

### ISSUE THREAD

**Issue by ewmailing** [pallene-lang/pallene#551]
This bug is discussed in https://github.com/pallene-lang/pallene/discussions/547
I have a reproducible case where something in Pallene seems to trigger Lua's internal LUAI_ASSERT.

This bug is a nightmare to isolate.
I have eliminated 99% of the program and hope this reproducible test case is now simple enough to use for diagonsing the problem.

Make sure to enable the LUAI_ASSERT if you want to trigger it.

- The original program used records, but to create this stripped down case, I made every thing use tables. Since I still hit the assertion, the problem is not related to records.
- The original program did tons of work to generate a data structure, which is fed into this final routine for processing, which is were the assertion failure happens. I have deleted all that code so there is no program, but I have serialized a dataset that would lead to the assertion failure. This new test case reloads all the data into a table so the problematic function can be called. My most trivial datasets were not able to trigger the assertion, so there is some particular amount of data needed, or a specific type to hit enough of the conditional cases in my function to trigger the assertion.
	- the serialized data is stored in debug_serialize_backTestPlot.lua (as Lua syntax)
- You need to build all the .pln files (ta, options, datetul, stocks_stackoverflow2)
	- ta and options modules are not actually used since I removed all the other code, but I didn't feeling like re-writing all the interfaces to finish removing them from the code.
- Removing too much code can actually cause the assertion to not trigger.
- Adding or moving code can also cause the assertion to happen somewhere else or not at all.
- The line it crashes at is a call into dateutil.IntegerToDate, but don't be fooled. This is called earlier, and was called a lot in my program. I also tried calling it a whole bunch of extra times earlier in this function, and it didn't do anything. So I don't think the routine itself is the problem. 
- Sometimes I was able to cause the assertion to happen earlier by declaring more local variables a little earlier and then moving the dateutil.DateToInteger call earlier right after those, e.g.
local tmp_arr1:{string}
local tmp_arr2:{string}
	I have removed a lot of code in the functions I provided you, so it may be a little harder to recreate this situation.

- The functions you care about are in stocks_stackoverflow2.pln.
	- GenerateBackTestPlotForChartJS
	- GenerateCodeForChartJSSeries

- These functions used to also call helpers in BackTesterChartJS_stackoverflow.lua, but I think I removed/disabled all the calls in this example.
- There are a few calls to regular Lua standard library functions such as concat and gsub. The external calls are prefixed with g_luaFuncs.

Build all the pln files. (The problem occurs with both -O0 and -O2. The problem does not occur in pure Lua via --emit-lua)
Then run:
lua pallene_stack_overflow_lua5.lua

I placed an io.write("About to crash here...") at the line the assertion fails for me.

I left some print statements to help track where the code goes. My console looks like this:
```
GenerateBackTestPlotForChartJS start
GenerateBackTestPlotForChartJS g_candleData
GenerateBackTestPlotForChartJS GenerateOptionsCode
GenerateBackTestPlotForChartJS plot_attributes_lines
GenerateBackTestPlotForChartJS inject
GenerateBackTestPlotForChartJS handle accessory
loop accessoryChartArray i:1
#accessory_plot.seriesArray:1
backTestPlot seriesArray: :1
Entered GenerateCodeForChartJSSeries
GenerateCodeForChartJSSeries second loop
second loop i::1
second loop y_array[i]::0.0
second loop x_array[i]::137869792378880
About to crash here...
lua: lapi.c:1015: lua_callk: Assertion `((nresults) == (0) || (L->ci->top - L->top >= (nresults) - (nargs))) && "results from function overflow current stack size"' failed.
Aborted (core dumped)
```
[stackoverflow.zip](https://github.com/pallene-lang/pallene/files/9619005/stackoverflow.zip)

**Comment by hugomg** [comment]
Thank you very much! A test case like this is extremely helpful. I can confirm that the assertion also triggers on my end and will try to debug it in the coming days when I find the time.

**Comment by ewmailing** [comment]
I was trying to workaround that bad function, and unfortunately after doing so, I hit the assertion failure in a completely different way in completely unrelated code. So I have a new second reproducible test case for you. (And I'm now kind of worried this bug is pervasive enough that there is no escape from it for me.)

I'm hoping between these two different examples, it will help make it easier to isolate the problem.

Compared to the previous program, this program has a smaller function that triggers the assert. There are also no string concats in this function, and only one for-loop.

I did not strip out all the dependencies in this one because the complicated dependencies are barely touched, and mostly just to get the initial data set to call the offending function. Once you get into that function with the data all loaded, the problems feel like they are happening within this function so the dependencies should not be getting in the way. Since the previous test case gutted out all the dependencies, I think it is established that all the other code is not a factor in this bug.

To further simplify things and demonstrate a point, I replaced a few function calls into a different module with a dummy re-implementation contained within the current module, and the bug still triggered. (You can see the commented out lines where the versions were swapped.)

I also added some comments in the code that show if you comment out certain lines of the code, the bug disappears.

I'm struggling to find a pattern...my best guess right now is that it requires a certain number of lines of code to trigger this bug. Removing too many things makes the bug go away.
Also, I noticed that if I removed some lines containing arrays, the bug sometimes goes away.


This program does have a couple of dependencies you need to install in order to run:
- lpeg
- lfs
- lua-curl

Build all the .pln files.
Then run:
lua pallene_stack_overflow2_lua.lua

The function to look at is m.on_init() in pallene_stack_overflow2_pln.pln. The assert gets triggered in the first iteration of the for-loop.
Take note of all the comments I left about how if you comment out different lines, the bug doesn't trigger.
I left print statements so you can see where/when the assert triggers.

My reference console output:
```
Downloading ticker data for: USO
Successfully downloaded data to: STOCK_CACHE/USO_1991-01-02_2022-08-12_1d.CSV
full_path_and_file	STOCK_CACHE/USO_1991-01-02_2022-08-12_1d.CSV
calling run
on_init
new_consensio
compute_position_percentages
create_transitions
get_chart
loop of #consensio_transitions: 80
loop index: 1
lua: lapi.c:1015: lua_callk: Assertion `((nresults) == (-1) || (L->ci->top - L->top >= (nresults) - (nargs))) && "results from function overflow current stack size"' failed.
Aborted (core dumped)
```

[stackoverflow2.zip](https://github.com/pallene-lang/pallene/files/9624082/stackoverflow2.zip)

**Comment by hugomg** [comment]
Maybe the reason the function must have enough lines is because larger functions need more stack space? If that's the case, it would be driven by the how many of the local variables contain a gc-able object (tables, strings, etc)

**Comment by hugomg** [comment]
When you tried that program with a lot of variables, did the function also use the variables or did it only declare them? Pallene does not allocate stack space for variables that are never used. Furthermore, if two variables are not used at the same time,  Pallene may have them share the same slot in the Lua stack. (The exact logic is over implemented in gc.lua).

**Comment by hugomg** [comment]
Quick way to test would be to add add a print statement in the gc.lua or the coder.lua to show the `max_frame_size` of the function in question.

**Comment by hugomg** [comment]
Yeah, I expect this function won't use a lot of stack space, because the variable lifetimes don't overlap. To maximize stack space, we want to move the variable uses all the way to the bottom of the function, because that way Pallene can't reuse stack slots.

**Comment by ewmailing** [comment]
Okay, I modified it to reuse variables at the bottom of the function, and I seem to have a reproducible case. Below is the general structure of the routine. Duplicate the lines to N. This should make a good test case since there are no dependencies and no data needed to setup up, unlike the prior two test cases.
 
```
function m.big_function(start_string:string, start_v1:integer, start_v2:integer) : string
	local big_arr:{string} = {}
	local v1 = start_v1
	local v2 = start_v2

	local v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)


	local str_array1:{string} = {}
	v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
	local tmp_str1:string = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
	str_array1[#str_array1+1] = tmp_str1 .. "\n"
	local ret_arr, ret_str = g_luaFuncs.back_in_lua(str_array1, tmp_str1)
	for i=1, #ret_arr do
		big_arr[#big_arr+1] = ret_arr[i]
		v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
		tmp_str1 = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
	end
	local str_array2:{string} = {}
	v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
	local tmp_str2:string = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
	str_array2[#str_array2+1] = tmp_str2 .. "\n"
	local ret_arr, ret_str = g_luaFuncs.back_in_lua(str_array2, tmp_str2)
	for i=1, #ret_arr do
		big_arr[#big_arr+1] = ret_arr[i]
		v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
		tmp_str2 = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
	end


	for i=1, #str_array1 do
		big_arr[#big_arr+1] = str_array1[i]
		big_arr[#big_arr+1] = tmp_str1
		v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
		tmp_str1 = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
		big_arr[#big_arr+1] = tmp_str1
	end
	for i=1, #str_array2 do
		big_arr[#big_arr+1] = str_array2[i]
		big_arr[#big_arr+1] = tmp_str2
		v1,v2,v3,v4,v5,v6,v7,v8,v9 = m.get_values(v1, v2)
		tmp_str2 = m.string_from_values(v1,v2,v3,v4,v5,v6,v7,v8,v9)
		big_arr[#big_arr+1] = tmp_str2
	end
	local flat_str = g_luaFuncs.concat(big_arr, '', 1, #big_arr)
	return flat_str
end
```
I started with 1024 repetitions and kept dividing by 2 until it stopped triggering the bug. At 32, it stopped triggering, so I went up to 48 and started working down again, but then I didn't hit the bug until around 8.

In my case, I usually was able to trigger the problem with 9 repetitions. A few times 8 repetitions triggered, but a few times it did not. Anyway, if you have problems, crank up the repetitions. I'm attaching my code generator with the zip.

Instructions:
```
lua print_lines.lua > pallene_stack_overflow_pln3b.pln
pallenec pallene_stack_overflow_pln3b.pln
lua pallene_stack_overflow_lua3b.lua
```

If you need more repetitions, change the local NUM_LOOPS = 9 to whatever number you want at the top of the print_lines.lua script.
I have compiled with both -O2 and -O0 optimizations levels, and both hit the assertion failure. For very large repetitions (e.g. 1024), I like to use -O0 because the compile time takes over a minute without optimizations, and longer with.


[stackoverflow3.zip](https://github.com/pallene-lang/pallene/files/9629769/stackoverflow3.zip)

**Comment by hugomg** [comment]
Aha! This info will help a lot.

**Comment by hugomg** [comment]
I might have found a simpler example, following that idea of a function with many local variables. Sometimes it segfaults and there was one time where it went in an infinite loop. BTW, the `.." "` bits are important because otherwise the strings would seen as compile time constants, which don't use extra stack space.

```lua
local M = {}

function M.foo(i: integer): string
    if i == 0 then
        return ""
    else
        local x01 = "a".." "
        local x02 = "b".." "
        local x03 = "c".." "
        local x04 = "d".." "
        local x05 = "e".." "
        local x06 = "f".." "
        local x07 = "g".." "
        local x08 = "h".." "
        local x09 = "i".." "
        local x10 = "j".." "
        local x11 = "k".." "
        local x12 = "l".." "
        local x13 = "m".." "
        local x14 = "n".." "
        local x15 = "o".." "
        local x16 = "p".." "
        local x17 = "q".." "
        local x18 = "r".." "
        local x19 = "s".." "
        local x20 = "t".." "
        local x21 = "u".." "
        local x22 = "v".." "
        local x23 = "w".." "
        local x24 = "x".." "
        local x25 = "y".." "
        local x26 = "z".." "

        local y = M.foo(i-1)

        local z = (
            x01..
            x02..
            x03..
            x04..
            x05..
            x06..
            x07..
            x08..
            x09..
            x10..
            x11..
            x12..
            x13..
            x14..
            x15..
            x16..
            x17..
            x18..
            x19..
            x20..
            x21..
            x22..
            x23..
            x24..
            x25..
            x26
        )

        return y .. z
    end
end

return M

```

**Comment by hugomg** [comment]
Hi @ewmailing, could you please check if PR #557 fixes that segmentation fault that you found? It fixes a nasty buffer overflow bug that happened because we were calculating the size of the Lua stack incorrectly. I can't understate how nasty of a bug this was. I'm honestly surprised that more things didn't break beforehand.

**Comment by hugomg** [comment]
Fixed by #557
